### PR TITLE
Reduce the number of intrinsic callables.

### PR DESCRIPTION
**Pull request by cgranade**
This PR replaces a number of intrinsic callables with pure-Q# implementations, reducing the size of the interface between the runtime and Q# applications. This PR is still in draft, as additional unit tests need to be added to ensure that the pure Q# implementations work as expected.

**Update:** added a bunch of tests, and updated to 0.15.

### CODE REVIEW

**Comment by bettinaheim** [comment]
@cgranade Would it make sense to add an implementation for the following: ModPowL, Round, PowD, and ExpD?

File: src/Simulation/QSharpFoundation/Math/Math.qs
Comment (irinayat-MS) [inline src/Simulation/QSharpFoundation/Math/Math.qs:100]: LLVM has intrinsic for log base 10 (https://llvm.org/docs/LangRef.html#id2078). Calling it _might_ be faster and have better precision than converting the bases ourselves.

File: src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs
Comment (irinayat-MS) [inline src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs:10]: Nit: because the test results are often reviewed by humans for both passing and failing tests on my previous projects we used to word the assert messages in terms of what is expected, then passing test ("green") would read as the expectation that was met and failing test ("red") would read as the expectation that wasn't met. It removed the cognitive dissonance when reviewing test logs with "Xor was incorrect." messages while still passing :) Instead, the message would be more like "Bitwise Xor of 11111000 and 00111111 should be 11000111"

Comment (cgranade) [reply src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs:10]: Thanks for raising that! In this case, `EqualityFactI` uses `FormattedFailure` to print the expected and actual values for each failing test, such that I think I the best way to incorporate that feedback is to add the arguments to the message:

```suggestion
        EqualityFactI(199, Xor(248, 63), "Xor(248, 63) was incorrect.");
```

File: src/Simulation/QSharpFoundation/Math/Math.qs
Comment (cgranade) [reply src/Simulation/QSharpFoundation/Math/Math.qs:100]: I'm fine either way; @bettinaheim had highlighted this as something that could be made concrete instead of intrinsic, but I'm happy to undo that change if that would be preferred.

File: src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs
Comment (irinayat-MS) [reply src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs:10]: My concern was mostly with having _negative_ message printed when the test is actually passing. Your call.

File: src/Simulation/QSharpFoundation/Math/Math.qs
Comment (bettinaheim) [reply src/Simulation/QSharpFoundation/Math/Math.qs:100]: There are also intrinsics for exp, sqrt and the like, but we first need to expose them via the .NET wrapper. I'd suggest to handle that here for now so that we can unblock e.g. the VQE example and start setting up more extensive testing, and then we can look into leveraging the llvm intrinsics directly.

File: src/Simulation/QSharpFoundation/Convert/Convert.qs
Comment (irinayat-MS) [inline src/Simulation/QSharpFoundation/Convert/Convert.qs:80]: @bettinaheim: Bettina, does this change mean that `%String* @__quantum__rt__double_to_string(double %val)` and other to string conversion functions can be removed from the QIR spec and from the runtime?

**Review by irinayat-MS — APPROVED** [review]

File: src/Simulation/QSharpFoundation/Convert/Convert.qs
Comment (bettinaheim) [reply src/Simulation/QSharpFoundation/Convert/Convert.qs:80]: Unfortunately, at least not yet. I don't have anything convenient to call in llvm directly and I would need to look in detail into how to do this using string initialization only. For now, this interpolated string will precisely result in a call to __quantum__rt__double_to_string and this change merely removed the redundancy in the QIR generation that would otherwise have required to handle the function call and the string interpolation separately.

File: src/Simulation/QSharpFoundation/Math/Math.qs
Comment (cgranade) [reply src/Simulation/QSharpFoundation/Math/Math.qs:100]: Sounds good, let me see if I can make some progress on `ModPowL`, `Round`, `PowD`, and `ExpD` as highlighted above, then. Thanks!

File: src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs
Comment (cgranade) [reply src/Simulation/Simulators.Tests/TestProjects/IntrinsicTests/Bitwise/Tests.qs:10]: Ah, got it; sorry for misunderstanding. Here, the string argument is the message to be printed in the case of failure, while the name of each test is taken from the name of the callable implementing the test. Thus here, `"Xor(248, 63) was incorrect."` only gets printed in the negative case, while `☑ XorIsCorrect` should be printed in the positive case.

File: src/Simulation/QSharpFoundation/Math/Math.qs
Comment (cgranade) [reply src/Simulation/QSharpFoundation/Math/Math.qs:140]: Those should now be added, resolving. Thank you!

**Comment by cgranade** [comment]
Since microsoft/QuantumLibraries#405 concerns the same part of the runtime, I also added an implementation of the `Microsoft.Quantum.Convert.DoubleAsInt` function from that proposal as a part of this PR. Thank you!

**Comment by cgranade** [comment]
It looks like there's a build failure when making `PI` a pure Q# function, so I backed out that change for now and will defer to a future PR once @swernli and I have had a chance to look at why only `PI` has to be intrinsic. That should allow for the build to complete and unblock you, @guenp.

**Comment by bettinaheim** [comment]
> Since [microsoft/QuantumLibraries#405](https://github.com/microsoft/QuantumLibraries/issues/405) concerns the same part of the runtime, I also added an implementation of the `Microsoft.Quantum.Convert.DoubleAsInt` function from that proposal as a part of this PR. Thank you!

I'll add support for that in the QIR emission with the next PR.

**Comment by swernli** [comment]
@cgranade I filed https://github.com/microsoft/qsharp-runtime/issues/502 to capture the issue with making `PI` intrinsic, and I intend to fix it as part of refactoring the test projects.

**Pull request merged**
